openedx_learning/core/publishing/model_mixins.py

```diff
@@ -0,0 +1,215 @@
+
+
+class PublishableEntityMixin(models.Model):
+    """
+    Convenience mixin to link your models against PublishableEntity.
+
+    Please see docstring for PublishableEntity for more details.
+
+    If you use this class, you *MUST* also use PublishableEntityVersionMixin and
+    the publishing app's api.register_content_models (see docstring for
+    details).
+
+    TODO: It would probably be useful to have a custom manager that
+    automatically pulls in publishing relations into the queryset.
+    """
+
+    publishable_entity = models.OneToOneField(
+        PublishableEntity, on_delete=models.CASCADE, primary_key=True
+    )
+
+    @cached_property
+    def versioning(self):
+        return self.VersioningHelper(self)
+
+    class Meta:
+        abstract = True
+
+    class VersioningHelper:
+        """
+        Helper class to link content models to their versions.
+
+        The publishing app has PublishableEntity and PublishableEntityVersion.
+        This is a helper class so that if you mix PublishableEntityMixin into
+        a content model like Component, then you can do something like::
+
+            component.versioning.draft  # current draft ComponentVersion
+            component.versioning.published  # current published ComponentVersion
+
+        It links the relationships between content models and their versioned
+        counterparts *through* the connection between PublishableEntity and
+        PublishableEntityVersion. So ``component.versioning.draft`` ends up
+        querying: Component -> PublishableEntity -> Draft ->
+        PublishableEntityVersion -> ComponentVersion. But the people writing
+        Component don't need to understand how the publishing models work to do
+        these common queries.
+
+        Caching Warning
+        ---------------
+        Note that because we're just using the underlying model's relations,
+        calling this a second time will returned the cached relation, and
+        not cause a fetch of new data from the database. So for instance, if
+        you do::
+
+            # Create a new Component + ComponentVersion
+            component, component_version = create_component_and_version(
+                learning_package_id=learning_package.id,
+                namespace="xblock.v1",
+                type="problem",
+                local_key="monty_hall",
+                title="Monty Hall Problem",
+                created=now,
+                created_by=None,
+            )
+
+            # This will work, because it's never been published
+            assert component.versioning.published is None
+
+            # Publishing happens
+            publish_all_drafts(learning_package.id, published_at=now)
+
+            # This will FAIL because it's going to use the relation value
+            # cached on component instead of going to the database again.
+            # You need to re-fetch the component for this to work.
```

Comment:
I'll add it, but basically you have to refetch it from the database. I'm hesitant to build anything smart around that because that's more of a Django ORM thing.

Reply:
@ormsbee Did you consider the collapsing `VersioningHelper` directly into the `PublishableEntityMixin` class?

In addition to simplifying the class structure, having `component.published` instead of `component.versioning.published` might be more immediately obvious to devs that `.published` will not automatically be updated without a re-fetch.

If I pretend I don't know this codebase, the `.versioning.` attribute in the middle kinda hints "magic happens here", even though the it's really just a syntactic helper construct.

Reply:
> @ormsbee Did you consider the collapsing `VersioningHelper` directly into the `PublishableEntityMixin` class?

I did. In fact, my original comment was proposing that. But I liked your suggestion of putting it behind `versioning` better, because I expect that a bunch of other stuff will make it into this class over time.

> In addition to simplifying the class structure, having `component.published` instead of `component.versioning.published` might be more immediately obvious to devs that `.published` will not automatically be updated without a re-fetch.
>
> If I pretend I don't know this codebase, the `.versioning.` attribute in the middle kinda hints "magic happens here", even though the it's really just a syntactic helper construct.

I mean, it is a _little_ magical, right? At least when compared to the simpler properties that just proxy to fields in the accompanying `publisher` app models (like `uuid`). But it's relatively small magic.

Maybe I should just change my implementation to always refetch. That's probably the more intuitive thing anyway, right?

Reply:
Gotcha. Sorry for the suggestion whiplash. When I suggested the `.versioning.` interface, I was thinking of it as an alternative to `PublishableEntity{Version}Mixin`. Since we're sticking with the mixins, I could go either way on `.versioning`. 

I think:
* `component.published` implies no re-fetch.
* `component.versioning.published` could go either way.
* `component.get_versioning().published` would heavily imply re-fetch.

Reply:
I'm going to stick with `versioning` for now. It is a bit magical, but I think it's more consistent with Django ORM conventions.

Reply:
👍🏻